src/main/java/org/springframework/data/couchbase/core/CouchbaseTemplate.java

```diff
@@ -148,28 +132,7 @@ public ReactiveCouchbaseTemplate reactive() {
 
 	@Override
 	public void setApplicationContext(final ApplicationContext applicationContext) throws BeansException {
-		prepareIndexCreator(applicationContext);
 		templateSupport.setApplicationContext(applicationContext);
 		reactiveCouchbaseTemplate.setApplicationContext(applicationContext);
 	}
-
-	private void prepareIndexCreator(final ApplicationContext context) {
-		String[] indexCreators = context.getBeanNamesForType(CouchbasePersistentEntityIndexCreator.class);
-
-		for (String creator : indexCreators) {
-			CouchbasePersistentEntityIndexCreator creatorBean = context.getBean(creator,
-					CouchbasePersistentEntityIndexCreator.class);
-			if (creatorBean.isIndexCreatorFor(mappingContext)) {
-				return;
-			}
-		}
-
-		if (context instanceof ConfigurableApplicationContext && indexCreator != null) {
-			((ConfigurableApplicationContext) context).addApplicationListener(indexCreator);
```

Comment:
So what are the actual use cases for these? I didn't see any tests for these use-cases/assumptions.

How would entities be added to the running application dynamically?

And for multiple MappingContext's they would need to create a corresponding new CouchbasePersistentEntityIndexCreator instance. Lots of places in the code make the assumption that there is only a single MappingContext.. for example the CouchbaseTemplate by way of CouchbaseConverter itself, so this isn't without precedence.

Reply:
If UserX is outside the @EnableCouchbaseRepositories package, its indexes won't be generated.
I don't know who would do this - but there's nothing preventing it.
```
        MappingContext mc= new CouchbaseMappingContext();
        CouchbaseConverter cvtr = new MappingCouchbaseConverter(mc);
        CouchbaseTemplate tmpl  = new CouchbaseTemplate(couchbaseClientFactory, cvtr);
        UserX u3 = tmpl.findById(UserX.class).one(user.get().getId());
```

Reply:
So it seems to me that this depends entirely on how the `MappingContext` was constructed.. it's typically the responsibility of the `@Configuration` bean to find and supply the context with potential entities..

In the vanilla Spring Data Couchbase, `AbstractCoucbbaseConfiguration#getInitialEntitySet()` uses `org.springframework.context.annotation.ClassPathScanningCandidateComponentProvider` scanning from the base package of the concrete class that extends `AbstractCoucbbaseConfiguration`.


In Spring Boot `org.springframework.boot.autoconfigure.data.couchbase.CouchbaseDataConfiguration` uses `org.springframework.boot.autoconfigure.domain.EntityScanner` which has its own logic for resolving potential entities.


And finally if someone is manually constructing beans as in your example, and if they forget to call `setInitialEntitySet()` then that's incorrect usage/a bug on their part?

Correct me if I'm wrong, but I don't believe that it's ever been the concern of the `MappingContext` to know **_how_** entities are found/resolved, that's always been something that's been externalised? Separation of concern..

Without test cases for the assumptions, it's really hard to ever know what the original intent was, and even harder to make sure we don't break it when making changes. But working through it logically I can't devine any meaningful reason that it needs to stick around? 

Especially since if you're going through the effort of creating multiple `MappingContext`'s and associated beans that depend on said `MappingContext`'s the functionality already exists to make it work as is.. as long as you glue things together correctly and call `setInitialEntitySet()` on the bespoke `MappingContext` instance?

@daschl's name is on the git blame for those original lines, perhaps he might have some more insight/context?

Reply:
"So it seems to me that this depends entirely on how the MappingContext was constructed."

It shouldn't and it doesn't.  But changing CouchbasePersistentIndexCreator  to a configuration bean would make it so.

"And finally if someone is manually constructing beans as in your example, and if they forget to call setInitialEntitySet() then that's incorrect usage/a bug on their part?"

No. Entities can be added afterwards, just as in the example I provided.  The applicationListener allows that.

"I don't believe that it's ever been the concern of the MappingContext to know how entities are found/resolved, that's always been something that's been externalised? "

It's (Abstract)MappingContext that resolves entity classes producing PersistentEntity objects. Maybe I misunderstood something.

Reply:
> It shouldn't and it doesn't. But changing CouchbasePersistentIndexCreator to a configuration bean would make it so.

So you're right in the sense that `AbstractMappingContext` will call `addPersistentEntity()` when `getPersistentEntity()` is called, which will emit a `MappingContextEvent`, when it encounters a previously unknown class. 

I think in my previous comment I was focusing too much on the `setInitialEntitySet()` call and I didn't think to much about the automatic resolution of unknown classes at runtime.

But because the example given was not using Spring, no `ApplicationContext` would be wired up via the `ApplicationContextAware` interface, and thus no event would have been emitted. So it still really does depend on how the `CouchbaseMappingContext` is constructed, manually vs Spring.

> No. Entities can be added afterwards, just as in the example I provided. The applicationListener allows that.

You're completely right about this one, and I missed it. Restoring the `ApplicationListener` interface to the `CouchbasePersistentIndexCreator` should resolve this.

But obviously, as above, only when Spring is involved.

The `initialEntitySet` will be used to create indices of classes known at `@Configuration` time and the `ApplicationListener` will handle the cases where classes are introduced into the mix at runtime.

> It's (Abstract)MappingContext that resolves entity classes producing PersistentEntity objects. Maybe I misunderstood something.

I'm thinking in the strict sense that it contains no logic to lookup, scan or resolve entities on it's own. One needs to supply/feed it by calling `setInitialEntitySet()` or at runtime by passing an unknown class to `addPersistentEntity()`. 
In that strict sense those are external actors providing it those classes.

We/I might be stuck on nomenclature of resolve vs scan here..

Reply:
> no ApplicationContext would be wired up via the ApplicationContextAware interface, and thus no event would have been emitted.

I edited the example to add the missing setApplicationContext() call. 

I'm not convinced that adding a second mechanism for auto-indexing is simpler than fixing the original mechanism.

Reply:
> I edited the example to add the missing setApplicationContext() call.

Actually it's needed on all the instances, both `MappingCouchbaseConverter` and `CouchbaseMappingContext` implement `ApplicationContextAware`. But at this point I feel the example is a little contrived, even more than you probably intended.

Suffice to say I can appreciate that not all entities will be discovered at `@Configuration` time, and some can come at runtime. Whether that is a bug or not can be argued about, but since `AbstractMappingContext` explicitly deals with it it make sense we conform as well.

> I'm not convinced that adding a second mechanism for auto-indexing is simpler than fixing the original mechanism.

The code/design in its current form is complex, it introduces unnecessary cognitive load for something that I've shown can be much simpler.

There is no need to pollute the `CouchbaseTemplate` class with concerns of glueing things together for the purpose of indexing.

The same goes for `CouchbaseMappingContext` class, it does not need to cross the boundary into the functionality of indexing. 

This doesn't need to be a special case:
https://github.com/spring-projects/spring-data-couchbase/blob/master/src/main/java/org/springframework/data/couchbase/core/mapping/CouchbaseMappingContext.java#L141-L147

There doesn't need to be a questionable cyclic dependency between `CouchbaseMappingContext` and `CouchbasePersistentEntityIndexCreator`.

Keeping things simple means having a single place that deals with indexing, and that seems to be the `CouchbasePersistentEntityIndexCreator`, it's right there in the name..

`CouchbasePersistentEntityIndexCreator` does not need to leak details of its internal implementation, and require logic in `CouchbaseMappingContext` and `CouchbaseTemplate` to make it work correctly.

I don't claim this is the simplest it can be, but it IS much simpler than the current design. 

127 lines removed and 44 added. 

Indexing logic in a single self-contained location.

Clean separation of concerns.

Reply:
Last commit causes this failure:

Error starting ApplicationContext. To display the conditions report re-run your application with 'debug' enabled.
2020-12-18 11:01:36.663 ERROR 17469 --- [           main] o.s.b.d.LoggingFailureAnalysisReporter   : 

***************************
APPLICATION FAILED TO START
***************************

Description:

The dependencies of some of the beans in the application context form a cycle:

   cmdRunner (field private com.test.UserRepository com.test.CmdRunner.userRepository)
      ↓
   couchbaseRepositoryOperationsMapping defined in class path resource [com/test/CouchbaseConfig.class]
┌─────┐
 |  couchbaseTemplate defined in class path resource [com/test/CouchbaseConfig.class]
↑                   ↓
 |  couchbaseMappingConverter defined in class path resource [com/test/CouchbaseConfig.class]
↑                    ↓
 |  couchbaseMappingContext defined in class path resource [com/test/CouchbaseConfig.class]
↑                    ↓
 |  couchbasePersistentEntityIndexCreator defined in class path resource [com/test/CouchbaseConfig.class]
└─────┘

Reply:
After fixing the cyclic dependency, classes found later do not get processed as the mappingContext they are emitted from is different from the the mappingContext of the CouchbasePersistentEntityIndexCreator.  This seems to be the reason that the indexCreator is tied to the template.

onApplicationEvent()
if (!enabled || !event.wasEmittedBy(mappingContext))

Reply:
@mikereiche crap, that'll serve me right for pushing a commit without running the build first.

fixed the cyclic dependency, and pushed.

I'm not seeing any more test failures now though

```
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 118, Failures: 0, Errors: 0, Skipped: 0
....
[INFO] 
[INFO] Tests run: 37, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
....
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  41.636 s
[INFO] Finished at: 2020-12-19T10:04:56+11:00
[INFO] ------------------------------------------------------------------------
```

Reply:
> emitted from is different from the the mappingContext of the CouchbasePersistentEntityIndexCreator. This seems to be the reason that the indexCreator is tied to the template.

So I've been thinking about this, I think I said it before but maybe not as succinctly. If you're explicitly creating a new `MappingContext`, then just like the need to explicitly create a dedicated `CouchbaseConverter` and `CouchbaseTemplate` that are bound to the new `MappingContext`, you should also create a dedicated `CouchbasePersistentEntityIndexCreator`. This seems to be the established pattern. 

It shouldn't be seen as a burden, for most people they won't even know about it since it's hidden away in a `@Configuration` class. And for those that need to customize, then it becomes another implementation detail they have to deal with.. like all the others.

But thinking a little bit more, we could however modify the `CouchbasePersistentEntityIndexCreator` to take a `List<MappingContext>` in the constructor instead of a single `MappingContext`, but I feel this goes against the already established pattern? 

And it won't work if you create things outside Spring, but I'm not very concerned with that. Because this is after all **_Spring_** Data Couchbase, so the expectation is that Spring is always involved to work it's magic, for better or worse.